staging/src/k8s.io/apiserver/pkg/registry/rest/delete_test.go

```diff
@@ -0,0 +1,245 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rest
+
+import (
+	"context"
+	"testing"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+type mockStrategy struct {
+	RESTDeleteStrategy
+	RESTGracefulDeleteStrategy
+}
+
+func (m mockStrategy) ObjectKinds(obj runtime.Object) ([]schema.GroupVersionKind, bool, error) {
+	gvk := obj.GetObjectKind().GroupVersionKind()
+	if len(gvk.Kind) == 0 {
+		return nil, false, runtime.NewMissingKindErr("object has no kind field ")
+	}
+	if len(gvk.Version) == 0 {
+		return nil, false, runtime.NewMissingVersionErr("object has no apiVersion field")
+	}
+	return []schema.GroupVersionKind{obj.GetObjectKind().GroupVersionKind()}, false, nil
+}
+
+func TestBeforeDelete(t *testing.T) {
+	type args struct {
+		strategy RESTDeleteStrategy
+		ctx      context.Context
+		obj      runtime.Object
+		options  *metav1.DeleteOptions
+	}
+
+	makePod := func(deletionGracePeriodSeconds int64) *v1.Pod {
+		return &v1.Pod{
+			TypeMeta: metav1.TypeMeta{APIVersion: "v1", Kind: "Pod"},
+			ObjectMeta: metav1.ObjectMeta{
+				DeletionTimestamp:          &metav1.Time{},
+				DeletionGracePeriodSeconds: &deletionGracePeriodSeconds,
+			},
+		}
+	}
+	makeOption := func(gracePeriodSeconds int64) *metav1.DeleteOptions {
+		return &metav1.DeleteOptions{
+			GracePeriodSeconds: &gracePeriodSeconds,
+		}
+	}
+
+	tests := []struct {
+		name                string
+		args                args
+		wantGraceful        bool
+		wantGracefulPending bool
+		wantErr             bool
+	}{
+		{
+			name: "when DeletionGracePeriodSeconds=-1, GracePeriodSeconds=-1",
+			args: args{
+				obj:     makePod(-1),
+				options: makeOption(-1),
+			},
+			wantGraceful:        false,
+			wantGracefulPending: true,
+		},
+		{
+			name: "when DeletionGracePeriodSeconds=-1, GracePeriodSeconds=0",
+			args: args{
+				obj:     makePod(-1),
+				options: makeOption(0),
+			},
+			wantGraceful:        false,
+			wantGracefulPending: true,
+		},
+		{
+			name: "when DeletionGracePeriodSeconds=-1, GracePeriodSeconds=1",
+			args: args{
+				obj:     makePod(-1),
+				options: makeOption(1),
+			},
+			wantGraceful:        false,
+			wantGracefulPending: true,
+		},
+		{
+			name: "when DeletionGracePeriodSeconds=-1, GracePeriodSeconds=2",
+			args: args{
+				obj:     makePod(-1),
+				options: makeOption(2),
+			},
+			wantGraceful:        false,
+			wantGracefulPending: true,
```

Comment:
I *think* I'd expect gracePeriodSeconds to be 1 after this call (-1 → 1, and is not allowed to be lengthened to 2)